heat/core/manipulations.py

```diff
@@ -1687,3 +1690,180 @@ def vstack(tup):
+                padding = torch.nn.ConstantPad1d(padding_sizes, neutral_value)
+                result = padding(result)
+                # Different value for indices padding to prevent type casting issues
+                padding = torch.nn.ConstantPad1d(padding_sizes, 0)
+                indices = padding(indices)
+        else:
+            result, indices = torch.topk(args[0], k=k, dim=dim, largest=largest, sorted=sorted)
+
+        # add offset of data chunks if reduction is computed across split axis
+        if dim == a.split:
+            offset, _, _ = a.comm.chunk(shape, a.split)
+            indices = indices.clone()
+            indices += torch.tensor(offset * a.comm.rank, dtype=indices.dtype)
+
+        local_shape = list(result.shape)
+        local_shape_len = len(shape)
+
+        metadata = torch.tensor([k, dim, largest, sorted, local_shape_len, *local_shape])
+        send_buffer = torch.cat(
+            (metadata.double(), result.double().flatten(), indices.flatten().double())
+        )
+
+        return send_buffer
+
+    gres = operations.__reduce_op(
+        a,
+        local_topk,
+        MPI_TOPK,
+        axis=dim,
+        neutral=neutral_value,
+        dim=dim,
+        sorted=sorted,
+        largest=largest,
+    )
+
+    # Split data again to return a tuple
+    local_result = gres._DNDarray__array
+    shape_len = int(local_result[4])
+
+    gres, gindices = local_result[5 + shape_len :].chunk(2)
+    gres = gres.reshape(*local_result[5 : 5 + shape_len].int())
+    gindices = gindices.reshape(*local_result[5 : 5 + shape_len].int())
+
+    # Fix the result in out to be a tuple
+
+    # Create output with correct split
+    if dim == a.split:
+        is_split = None
+        split = a.split
+    else:
+        is_split = a.split
+        split = None
+
+    final_array = factories.array(
+        gres, dtype=a.dtype, device=a.device, split=split, is_split=is_split
+    )
+    final_indices = factories.array(
+        gindices, dtype=torch.int64, device=a.device, split=split, is_split=is_split
+    )
+
+    if out is not None:
+        if out[0].shape != final_array.shape or out[1].shape != final_indices.shape:
+            raise ValueError(
+                "Expecting output buffer tuple of shape ({}, {}), got ({}, {})".format(
+                    gres.shape, gindices.shape, out[0].shape, out[1].shape
+                )
+            )
+        out[0]._DNDarray__array.storage().copy_(final_array._DNDarray__array.storage())
+        out[1]._DNDarray__array.storage().copy_(final_indices._DNDarray__array.storage())
+
+        out[0]._DNDarray__dtype = a.dtype
+        out[1]._DNDarray__dtype = types.int64
+
+    return final_array, final_indices
+
+
+def mpi_topk(a, b, mpi_type):
+    # Parse Buffer
+    a_parsed = torch.from_numpy(np.frombuffer(a, dtype=np.float64))
+    b_parsed = torch.from_numpy(np.frombuffer(b, dtype=np.float64))
+
+    # Collect metadata from Buffer
+    k = int(a_parsed[0].item())
```

Comment:
parsing necessary? can the values be left as float64 instead?

Reply:
No, torch.topk only accepts int/bool values. Throws TypError otherwise.